### ISSUE TITLE
DDG4: G4MTRunManager help

### ISSUE THREAD

**Issue by decibelcooper** [AIDASoft/DD4hep#372]
Hi All.  What is the status of the G4MTRunManager plugin?  Is it fully functional?  I can't seem to get past an error once my workers are set up:

> **************************************************** 
> *  A runtime error has occured :                     
> *    GeneratePrimaries: Panic! No action sequencer defined. No primary particles can be produced.
> *  the program will have to be terminated - sorry.   
> **************************************************** 

I am using python for setup.  I have callbacks for setting up worker, master, sensitives, and fields.  I call the kernel's generatorAction() for the sequence, and adopt my actions in the same way that I did with G4RunManager, except that I've tried to do this both in the worker setup callback, as well as in the main thread.  I'm trying to follow section 7 of the DDG4 manual, but without success.

**Comment by MarkusFrankATcernch** [comment]
I checked that the multi-threaded execution mode still works.
Please see the pull request #374 for details. There are some caveats though:
-- Geant4 must be build with multi-threading support. By default this option is not enabled.
-- The plugin name "G4MTRunManager" must be set to the Geant4Kernel as a property.
    (If someone has a clever policy when to set it automatically this perhaps could be omitted).
I did not see any problems running a couple of events in parallel.
If all race conditions are handled properly must still be verified with longer runs - I did my best at the design level, but the ultimate proof is only be experience.

**Comment by decibelcooper** [comment]
Hi @MarkusFrankATcernch 

Thank you for your reply and your update to the SiD examples!  It has helped me to know that the problem is on my end, so that I know where to look for the problem (thought I'd take another crack at it before using up more of your time).  I have solved it, and it was an embarrassingly simple problem.  In the worker-specific callback, I was building the action sequences in the master context rather than the specific worker context.

The reason for this mistake might be worth mentioning.  I only ran into this trouble because I refuse to use the `Geant4` native python class defined at https://github.com/AIDASoft/DD4hep/blob/cc7b5b5fa8fdec55171eccce191daf27688a11ed/DDG4/python/DDG4.py#L327.  This is because I believe that it obscures the underlying API too much, and makes it difficult for a user to understand what is being done in the setup (which is very important to me, when significant man and CPU cycles are put into the simulations).  I understand the desire to factorize the boiler-plate code into the python module, but If I were to make a suggestion from this experience, I would say that the examples should not rely on this factorization.  I.e., please provide examples that use the API more directly and allow a better understanding for the user.

That being said, now that I understand some of these MT-related aspects of the code better, I do quite appreciate it.  It is very nice work!  I have just one potential bug fix to request, but I will create a new Issue and a PR for your consideration.
